feat: instrument DataDog

zchsh opened the pull request:
- feat: add DataDog, via ConsentManager
- perf: prefetch DataDog script

## 🔗 Relevant links

<!--
Include links to the branch preview, Asana task, and Figma designs wherever possible to make reviewing your PR easier.
When including the preview link, make sure to remove any '.' characters from the branch name:
  - Example: ks.my-branch -> ksmy-branch
-->

- [Preview link](https://dev-portal-git-BRANCH_NAME-hashicorp.vercel.app/PATH_TO_VIEW) 🔎
- [Asana task](url) 🎟️

## 🗒️ What

<!--
Briefly list out the changes proposed in this PR.
-->

## 🤷 Why

<!--
Describe why the changes proposed are needed. Some examples: new feature requested, refactor to make things easier later, styling tweaks requested by design, etc.
-->

## 🛠️ How

<!--
Dive into the approach you took, list resources you referenced, detail other approaches you tried but didn't end up going with, etc.
-->

## 📸 Design Screenshots

<!--
Include a screenshot or two and a link to the designs you referenced with this code. These are both helpful context for reviewers to understand what designs you were looking at when you were putting this code together.
-->

## 🧪 Testing

<!--
Create a checklist for going through how to test your proposed changes. If there is anything to configure before interacting with the project in a browser, such as toggling feature flags, changing machine settings, or simulating behavior in browser dev tools, list those steps first.

- [ ] Step 1
- [ ] Step 2
- [ ] Step 3
- [ ] ...
-->

## 💭 Anything else?

<!--
If there is anything you came across that you chose not to address in this PR but plan to soon, list those items here and any Asana tasks you created to go with them.
-->

zchsh commented on src/layouts/base-new/index.tsx:
```diff
@@ -27,6 +33,9 @@ const BaseNewLayout: React.FC<BaseNewLayoutProps> = ({
   return (
     <>
+      <Head>
+        <link rel="prefetch" href={DATADOG_SCRIPT_URL} />
```
Intent here is to prefetch the DataDog RUM script so that, if consent is given, the script will load more quickly.

zchsh commented on src/lib/consent-manager-services/dev-dot.ts:
```diff
@@ -0,0 +1,48 @@
+      applicationId: '${DATADOG_APP_ID}',
+      site: '${DATADOG_SITE}',
+      env: '${ENV}', 
+      trackInteractions: false,
```
Wanted to check on whether we want to enable `trackInteractions`. DataDog's snippets show it set to `true`, however they also explicitly note:

> The trackInteractions parameter enables the automatic collection of user clicks in your application. Sensitive and private data contained on your pages may be included to identify the elements interacted with.

I've set to `false` for now as I'm not 100% sure on the potential for sensitive and private data with dev-portal.

mhonnatti replied:
I'm okay with starting this way! We'll need to enable it eventually for session replays

BRKalow commented:
I believe we're currently working to get DD to stop collecting IP addresses, which would allow us to not have to implement this behind consent manager. In that case, maybe we remove the consent manager gate here in preparation for that?

I think we'll also want to set the `sampleRate` fairly low. Maybe not yet as we don't have real traffic, but definitely before the launch. (https://docs.datadoghq.com/real_user_monitoring/browser/#configuration)

mhonnatti commented:
@zchsh @BRKalow confirming that DataDog needs to be outside the consent manager.

mhonnatti commented on src/lib/datadog/index.tsx:
```diff
@@ -0,0 +1,54 @@
+
+function DataDogScriptTag() {
+  return (
+    <Script id={dataDogService.name} strategy="afterInteractive">
```
considering that we will be using Datadog for web vitals and console error logging, should this be `beforeInteractive` instead of `afterInteractive`? @BRKalow thoughts?

BRKalow replied:
@mhonnatti Yeah, that seems reasonable for now. We'll need to shift this to `_document` then, per https://nextjs.org/docs/basic-features/script#beforeinteractive

mhonnatti commented:
@zchsh since you still don't have Datadog access, here's the snippet that Datadog generated for us:

`
import { datadogRum } from '@datadog/browser-rum';

datadogRum.init({
    applicationId: '4486817b-44b8-4e6c-b6a4-cef83e48cf46',
    clientToken: 'pub750ceadc5a9c79a6d3da59b534eb5108',
    site: 'datadoghq.com',
    service:'developer.hashicorp.com',
    env:'env',
    // Specify a version number to identify the deployed version of your application in Datadog 
    // version: '1.0.0',
    sampleRate: 100,
    premiumSampleRate: 100,
    trackInteractions: true,
    defaultPrivacyLevel:'mask-user-input'
});
    
datadogRum.startSessionReplayRecording();
`
We've already addressed `trackInteractions`. `site` looks like it needs to be updated, `premiumSampleRate` needs to be added

zchsh commented:
@mhonnatti @BRKalow I've updated this PR with config values (thanks Manish!), and have made the following changes as well:
- Decoupled from `ConsentManager`, now loads script directly via `next/script`
- Moved to `_document.tsx`, and changed `strategy` to `beforeInteractive`
- Added `version` config, set to the commit SHA from Vercel
- For `sampleRate`, have left at `100` for now for testing purposes. Happy to adjust before merging if there's a specific number we want (eg `10`?). But figure we might also be fine to leave this for a follow-up PR before launch (after I can more thoroughly confirm this is working by logging into Datadog)

Thanks for both of your help here, and let me know what ya'll think 🙇

zchsh commented on src/lib/datadog/index.tsx:
```diff
@@ -0,0 +1,48 @@
+      trackInteractions: false,
+      defaultPrivacyLevel:'mask-user-input'
+    })
+    DD_RUM.startSessionReplayRecording()
```
I'm unclear what this will do with `trackInteractions: false`, but figured might as well add it in so that we don't have to remember to do so if / when we set `trackInteractions: true`

zchsh commented:
@mhonnatti @BRKalow apologies for the delay here, this should be ready for re-review! Have updated as follows:

- Sticking with `afterInteractive`
- Added variable config per environment, and using a separate Datadog app for prod vs non-prod
- Moved script tag to `dev-dot` specific layout file, so that this doesn't run on `dot-io`s

mhonnatti reviewed (approved):
🤘
